libmemcached/tls.c

```diff
@@ -0,0 +1,595 @@
+    memcached_return_t rc;
+    if (*want == WANT_READ) {
+        rc = memcached_io_wait_for_read(instance);
+    } else if (*want == WANT_WRITE){
+        rc = memcached_io_wait_for_write(instance);
+    } else {
+        // no error
+        return false;
+    }
+
+    if (memcached_success(rc)) {
+        return true;
+    }
+
+    return false;
+}
+
+ssize_t memcached_ssl_read(memcached_instance_st* instance,
+             char* input_buf,
+             size_t buffer_length,
+             int flags){
+    ssize_t rv = 0;
+    memcached_SSL *memc_ssl = (memcached_SSL*)instance->privctx;
+    if (!memc_ssl) {
+        memcached_set_error(*instance, MEMCACHED_TLS_ERROR, MEMCACHED_AT, memcached_literal_param("Couldn't execute SSL read command, SSL isn't initialized"));
+        return -1;
+    }
+    SSL *ssl = memc_ssl->ssl;
+
+    do {
+        want_t want = WANT_NONE;
+        int nread = SSL_read(ssl, input_buf, buffer_length);
+        rv = handle_ssl_return_value(instance, nread, &want);
+        if (want != WANT_NONE && handle_renegotation_error(instance, &want)) {
+            // WANT error handled, try again
+            continue;
+        } else {
+            break;
+        }
+    } while(false);
+    return rv;
+}
+
+ssize_t memcached_ssl_write(memcached_instance_st* instance,
+             char* local_write_ptr,
+             size_t write_length,
+             int flags){
+    ssize_t rv = 0;
+    memcached_SSL *memc_ssl = (memcached_SSL*)instance->privctx;
+    if (!memc_ssl) {
+        memcached_set_error(*instance, MEMCACHED_TLS_ERROR, MEMCACHED_AT, memcached_literal_param("Couldn't execute SSL write command, SSL isn't initialized"));
+        return -1;
+    }
+    SSL *ssl = memc_ssl->ssl;
+
+    do {
+        want_t want = WANT_NONE;
+        int nread = SSL_write(ssl, local_write_ptr, write_length);
+        rv = handle_ssl_return_value(instance, nread, &want);
+        if (want != WANT_NONE && handle_renegotation_error(instance, &want)) {
+            // WANT error handled, try again
+            continue;
+        } else {
+            break;
+        }
+    } while(false);
+    return rv;
+}
+
+memcached_return_t memcached_ssl_get_server_certs(memcached_instance_st * instance, char *output)
+{
+    SSL* ssl;
+    X509 *cert;
+    char *line;
+    if (instance->privctx == NULL) {
+        return memcached_set_error(*instance, MEMCACHED_TLS_ERROR, MEMCACHED_AT, memcached_literal_param("The instance doesn't have SSL context\n"));
+    }
+    ssl = ((memcached_SSL *)instance->privctx)->ssl;
+    if (ssl == NULL) {
+        return memcached_set_error(*instance, MEMCACHED_TLS_ERROR, MEMCACHED_AT, memcached_literal_param("SSL connection wasn't yet established for this server\n"));
+    }
+    cert = SSL_get_peer_certificate(ssl); /* get the server's certificate */
+    if ( cert != NULL )
+    {
+        char * certs_msg = (char *)libmemcached_malloc(instance->root, 100);
```

Comment:
Check that you are freeing this properly from the caller to avoid memory leak.

Reply:
I'm not using this function inside the code, it is supposed to be for the endpoint user use. 
I added to it's description:
 * The function allocates memory Using memcached_st's malloc function and "output" should be freed by the user.